### PR TITLE
rustc_metadata: Switch module children decoding to an iterator

### PR DESCRIPTION
**Pull request by petrochenkov**
Previously https://github.com/rust-lang/rust/pull/103578, https://github.com/rust-lang/rust/pull/103524 and previous PRs simplified it as much as possible.

A couple of cleanup commits is also added.
r? @cjgillot

### CODE REVIEW

**Review by cjgillot — APPROVED** [review]
2 questions, and r=me

File: compiler/rustc_resolve/src/build_reduced_graph.rs [hunk 1]
```diff
@@ -204,7 +204,9 @@ impl<'a> Resolver<'a> {
     }
 
     pub(crate) fn build_reduced_graph_external(&mut self, module: Module<'a>) {
-        for child in self.cstore().module_children_untracked(module.def_id(), self.session) {
+        for child in
+            Vec::from_iter(self.cstore().module_children_untracked(module.def_id(), self.session))
```
[the remark this answers is not in the document]
Comment (petrochenkov) [reply compiler/rustc_resolve/src/build_reduced_graph.rs:208]: Due to borrow checker, `self` is mutably borrowed inside the loop.

File: compiler/rustc_metadata/src/rmeta/encoder.rs [hunk 2]
```diff
@@ -1305,6 +1298,11 @@ impl<'a, 'tcx> EncodeContext<'a, 'tcx> {
                     }
                 }
             }));
+
+            if let Some(reexports) = tcx.module_reexports(local_def_id) {
+                assert!(!reexports.is_empty());
```
[the remark this answers is not in the document]
Comment (petrochenkov) [reply compiler/rustc_metadata/src/rmeta/encoder.rs:1303]: This assert is very cheap (both in absolute value, and relatively to everything else happening here), so I think we can always enable it.

**Pull request merged**